### ISSUE TITLE
Generic descriptors on generic base classes produce an unknown type in subclasses

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#2346]
**Describe the bug**
Given the following descriptor and decorator:

```python
from __future__ import annotations

from collections.abc import Callable
from typing import Any, Generic, TypeVar, overload

T_contra = TypeVar('T_contra', contravariant=True)
V_co = TypeVar('V_co', covariant=True)
CachedSlotPropertyT = TypeVar(
    'CachedSlotPropertyT', bound='CachedSlotProperty[Any, Any]'
)


class CachedSlotProperty(Generic[T_contra, V_co]):
    def __init__(self, name: str, function: Callable[[T_contra], V_co], /) -> None:
        self.name: str = name
        self.function: Callable[[T_contra], V_co] = function
        self.__doc__: str | None = function.__doc__

    @overload
    def __get__(
        self: CachedSlotPropertyT, instance: None, owner: type[T_contra]
    ) -> CachedSlotPropertyT:
        ...

    @overload
    def __get__(self, instance: T_contra, owner: Any) -> V_co:
        ...

    def __get__(
        self: CachedSlotPropertyT, instance: T_contra | None, owner: Any
    ) -> CachedSlotPropertyT | V_co:
        if instance is None:
            return self

        try:
            return getattr(instance, self.name)
        except AttributeError:
            value = self.function(instance)
            setattr(instance, self.name, value)
            return value


def cached_slot_property(
    name: str,
    /,
) -> Callable[[Callable[[T_contra], V_co]], CachedSlotProperty[T_contra, V_co]]:
    def decorator(
        func: Callable[[T_contra], V_co],
        /,
    ) -> CachedSlotProperty[T_contra, V_co]:
        return CachedSlotProperty(name, func)

    return decorator
```

Non-generic classes and subclasses get inferred correctly:

```python
class A:
    __slots__ = ('_prop',)

    @cached_slot_property('_prop')
    def prop(self) -> int:
        return 0


class B(A):
    ...


A.prop  # CachedSlotProperty[A, int]
B.prop  # CachedSlotProperty[A, int]

a = A()
a.prop  # int

b = B()
b.prop  # int
```

But subclasses of a generic base do not:

```python
T = TypeVar('T')


class C(Generic[T]):
    __slots__ = ('_data', '_prop')

    def __init__(self, data: T) -> None:
        self._data: T = data

    @cached_slot_property('_prop')
    def prop(self) -> int:
        return 0


class D(C[int]):
    ...


C.prop  # CachedSlotProperty[Any, int]
D.prop  # CachedSlotProperty[Any, int]

c = C('asdf')
c.prop  # Unknown

d = D(1)
d.prop  # Unknown
```

**Expected behavior**
`C.prop` would be inferred as `CachedSlotProperty[C[T@C], int]`, `D.prop` would be inferred as `CachedSlotProperty[C[int], int]`, and `c.prop` and `d.prop` would be inferred as `int`.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line

Which version?
1.1.171

**Additional context**
If the signature of `CachedSlotProperty.__get__` is changed to use `Any` in place of `T_contra`, the property types infer correctly, but this still doesn't fix the inference of the class-level descriptor overload.

**Comment by erictraut** [comment]
Thanks for the bug report.

On initial investigation, the observed behavior is related to the way that pyright handles unannotated `self` and `cls` parameters. When a `self` parameter is unannotated, pyright assumes its type is a type variable that is bound to the containing class's type. It's as though you had explicitly specified the following:

```python
__type_of_self_C = TypeVar("__type_of_self_C", bound="C[Any]")

class C(Generic[T]):
    @cached_slot_property("_prop")
    def prop(self: __type_of_self_C) -> int:
        return 0
```

If I explicitly annotate `self` as `C[T]`, the error is eliminated, but the type of `d.prop` is not what you would expect.

I'll need to investigate further.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.173, which I just published. It will also be included in the next release of pylance.
